modules/diskq/diskq-serializer.c

```diff
@@ -0,0 +1,195 @@
+}
+
+static DiskQSerializerCache *
+_cache_new(gsize chunk_size)
+{
+  DiskQSerializerCache *self = g_new0(DiskQSerializerCache, 1);
+  self->buffer = g_string_sized_new(chunk_size);
+  self->pos = 0;
+
+  return self;
+}
+
+static void
+_cache_free(DiskQSerializerCache *self)
+{
+  g_string_free(self->buffer, TRUE);
+  g_free(self);
+}
+
+static gboolean
+_cache_fetch(DiskQSerializerCache *self, gchar *buf, gsize buflen)
+{
+  if (self->pos + buflen > self->buffer->len)
+    return FALSE;
+
+  memcpy(buf, &self->buffer->str[self->pos], buflen);
+  self->pos += buflen;
+  return TRUE;
+}
+
+static void
+_cache_move_data_to_front(DiskQSerializerCache *self)
+{
+  gsize data_len = self->buffer->len - self->pos;
+  memmove(self->buffer->str, &self->buffer->str[self->pos], data_len);
+  self->pos = 0;
+  g_string_set_size(self->buffer, data_len);
+}
+
+static gsize
+_cache_allocated_size(DiskQSerializerCache *self)
+{
+  return self->buffer->allocated_len;
+}
+
+static gsize
+_cache_data_size(DiskQSerializerCache *self)
+{
+  return (self->buffer->len - self->pos);
+}
+
+static void
+_cache_increase_buffer(DiskQSerializerCache *self, gsize buflen)
+{
+  gsize orig_len = self->buffer->len;
+  g_string_set_size(self->buffer, self->buffer->allocated_len + buflen);
+  g_string_set_size(self->buffer, orig_len);
+}
+
+static gchar *
+_cache_get_buffer_end(DiskQSerializerCache *self)
+{
+  return &self->buffer->str[self->buffer->len];
+}
+
+static void
+_cache_increase_data_size(DiskQSerializerCache *self, gsize buflen)
+{
+  g_string_set_size(self->buffer, self->buffer->len + buflen);
+}
+
+static gboolean
+_push_data_to_cache(MsgSerializer *self, gsize buflen)
+{
+  gsize chunk_size = MAX(buflen, self->chunk_size);
+
+  if ((self->cache->pos % 16) == 0)
+    _cache_move_data_to_front(self->cache);
+
+  if (_cache_data_size(self->cache) + chunk_size >= _cache_allocated_size(self->cache))
+    _cache_increase_buffer(self->cache, chunk_size);
+
+  ssize_t retval = _pread_all(self->fd, _cache_get_buffer_end(self->cache), chunk_size, self->offset);
+  if (retval <= 0)
+    return FALSE;
+
+  _cache_increase_data_size(self->cache, chunk_size);
+
+  self->offset += retval;
+  return TRUE;
+}
+
+gboolean
+msg_serializer_read_bytes(SerializeArchive *s, gchar *buf, gsize buflen, GError **error)
+{
+  MsgSerializer *self = (MsgSerializer *) s;
+
+  g_return_val_if_fail(error == NULL || (*error) == NULL, FALSE);
+
+  if (buflen == 0)
+    return TRUE;
+
+  if (_cache_fetch(self->cache, buf, buflen))
+    return TRUE;
+
+  if (!_push_data_to_cache(self, buflen))
+    return FALSE;
+
+  g_assert(_cache_fetch(self->cache, buf, buflen));
+
+  return TRUE;
+}
+
+static void
+msg_serializer_free(SerializeArchive *s)
+{
+  MsgSerializer *self = (MsgSerializer *)s;
+  _cache_free(self->cache);
+}
+
+SerializeArchive *
+msg_serializer_new_full(int fd, gsize offset, gsize chunk_size)
+{
+  MsgSerializer *self = g_slice_new0(MsgSerializer);
```

Comment:
why are you using `g_slice`?

Reply:
This is unfortunate thing. The free function uses this logic too.

```
void
serialize_archive_free(SerializeArchive *self)
{
  g_clear_error(&self->error);
  g_slice_free1(self->len, self);
}
```

I do not know any reason why would we want to use slices. On the other hand, if I want to change it, I need to change all of the original serializers, which is not too much code, but I do not want to open the scope with it right now.